design/open/uncomitted-gc.md

```diff
@@ -0,0 +1,119 @@
+   1. Lexicographically sortable, descending time based serialization
+   2. A unique identifier for the lakeFS instance
+      `<sortable_descending_serialized_uid>_<lakefs_instance_uid>`  
+This serialized partition prefix will allow partial scans of the bucket when running the optimized GC
+4. The lakeFS instance will track the count of objects uploaded to the prefix and create a new partition every < TBD > objects uploaded
+
+The full object path for an object `file1` uploaded to repository `my-repo` on branch `test-branch` will be as follows:  
+    
+    <lakeFS_bucket>/repos/my-repo/branches/test-branch/<lakeFS_instance_partition_uid>/<file1_uid>
+
+#### StageObject
+
+1. Allowed only for address outside the repo namespace
+2. Prevent race between staging object and GC job
+
+#### Get\Link PhysicalAddress 
+
+1. GetPhysicalAddress to return a validation token along with the address.
+2. The token will be valid for a specified amount of time and for a single use.
+3. LinkPhysicalAddress to verify token valid before creating an entry.
+    1. Doing so will allow us to use this time interval to filter objects that might have been uploaded and waiting for
+       the link API and avoid them being deleted by the GC process.
+    2. Objects that were uploaded to a physical address issued by the API and were not linked before the token expired will
+       eventually be deleted by the GC job.
+
+#### CopyObject
+
+1. Copy object in the same branch will work the same - creating a new staging entry using the existing entry information.
+2. For objects that are not part of the branch, use the underlying adapter copy operation.
+3. "Move/Rename" operation will need to use the new version of CopyObject instead of StageObject.
+
+With the above changes, bucket data is partitioned by repositories, branches and finally a bounded partition.
+This allows running the GC process at branch level, reduces the data scan size and allows parallelization.
+
+#### PrepareUncommittedForGC
+
+A new API which will create meta-ranges and ranges using the given branch uncommitted data. These files
+will be saved in a designated path used by the GC client to list branch's uncommitted objects.  
+For the purpose of this document we'll call this the `GC commit` (feel free to suggest a better term :) )
+
+#### Reading data from lakeFS
+
+Reading data from lakeFS will be similar to the current GC process - using the SSTable reader to quickly list branch objects.
+The above-mentioned API will enable doing so for all branch objects (committed and uncommitted).
+
+### GC Flows
+
+The following describe the GC process run flows in the branch scope:
+
+#### Flow 1: Clean Run
+
+1. Run PrepareUncommittedForGC
+2. Read all addresses from branch commits -> `lakeFS DF`
+3. Read all addresses from branch `GC commit` -> `lakeFS DF`
+4. Read all objects on branch path directly from object store (can be done in parallel by 'partition') -> `Branch DF`
```

Comment:
Is this a listing on the object store? In what sense is it parallel?

Reply:
Since the objects on a given branch a divided into partition prefixes we can use workers to list objects by partition and aggregate the result